fw/http.c

```diff
@@ -5709,16 +5902,61 @@ tfw_http_req_process(TfwConn *conn, TfwStream *stream, struct sk_buff *skb)
 				__set_bit(TFW_HTTP_B_HEADERS_PARSED, req->flags);
 			}
 
-			if (tfw_h2_stream_req_complete(req->stream)) {
-				if (likely(!tfw_h2_parse_req_finish(req)))
-					break;
+			if (!test_bit(TFW_HTTP_B_HEADERS_PARSED, req->flags))
+				goto skip;
+
+			if (ctx->hdr.type == HTTP2_DATA) {
+				if (stream->proto == HTTP2_STREAM_PROTO_WEBSOCKET)
+					return tfw_ws_msg_process(conn, stream,
+								  skb);
+			}
+
+			if (!test_bit(TFW_HTTP_B_UPGRADE_WEBSOCKET, req->flags)) {
+				if (unlikely(
+					!TFW_STR_EMPTY(
+						&ht->tbl[TFW_HTTP_HDR_H2_PROTOCOL])
+					&& req->method == TFW_HTTP_METH_CONNECT))
+				{
+					__h2_msg_hdr_val(
+						&ht->tbl[TFW_HTTP_HDR_H2_PROTOCOL],
+						&protocol_val);
+					if (!tfw_str_eq_cstr(&protocol_val, S_WEBSOCKET,
+							SLEN(S_WEBSOCKET),
+							TFW_STR_EQ_CASEI))
+					{
+						T_WARN("Unsupported upgrade protocol\n");
+						return TFW_BLOCK;
+					}
+					__set_bit(TFW_HTTP_B_UPGRADE_WEBSOCKET,
+						req->flags);
+				}
+			}
+
+			if (test_bit(TFW_HTTP_B_UPGRADE_WEBSOCKET, req->flags)
+			    || tfw_h2_stream_req_complete(req->stream))
+			{
+				if (test_bit(TFW_HTTP_B_UPGRADE_WEBSOCKET,
+					     req->flags))
+				{
+					if (likely(
+						!tfw_h2_parse_req_semifinish(req)))
+					{
+						break;
+					}
```

Comment:
It really can be cumbersome to move code into separate function. Because we need here differentiate four code paths: normal flow, break from switch, skip to label, return from function. That is not practical here.

Reply:
Braces fixed.